OpenRA.Mods.Common/UpdateRules/Rules/DefineLocomotors.cs

```diff
@@ -0,0 +1,164 @@
+					"and affected Mobile traits are listed for inspection.";
+			}
+		}
+
+		readonly List<Tuple<string, string>> locations = new List<Tuple<string, string>>();
+		bool subterraneanUsed;
+		bool jumpjetUsed;
+
+		readonly string[] locomotorFields =
+		{
+			"TerrainSpeeds", "Crushes", "CrushDamageTypes", "SharesCell", "MoveIntoShroud"
+		};
+
+		readonly string[] subterraneanFields =
+		{
+			"SubterraneanTransitionCost", "SubterraneanTransitionTerrainTypes",
+			"SubterraneanTransitionOnRamps", "SubterraneanTransitionDepth",
+			"SubterraneanTransitionPalette", "SubterraneanTransitionSound",
+		};
+
+		readonly string[] jumpjetFields =
+		{
+			"JumpjetTransitionCost", "JumpjetTransitionTerrainTypes", "JumpjetTransitionOnRamps"
+		};
+
+		public override IEnumerable<string> AfterUpdate(ModData modData)
+		{
+			var message = "You must define a set of Locomotor traits to the World actor for the different\n"
+				+ "movement classes used in your mod (e.g. Infantry, Vehicles, Tanks, Ships, etc)\n"
+				+ "and replace any definitions/overrides of the following properties on each\n"
+				+ "actor with a Locomotor field referencing the appropriate locomotor type.\n\n"
+				+ "The standard Locomotor definition contains the following fields:\n"
+				+ UpdateUtils.FormatMessageList(locomotorFields) + "\n\n";
+
+			if (subterraneanUsed)
+				message += "Actors using the subterranean logic should reference a SubterraneanLocomotor\n"
+				+ "instance that extends Locomotor with additional fields:\n"
+				+ UpdateUtils.FormatMessageList(subterraneanFields) + "\n\n";
+
+			if (jumpjetUsed)
+				message += "Actors using the jump-jet logic should reference a JumpjetLocomotor\n"
+				+ "instance that extends Locomotor with additional fields:\n"
+				+ UpdateUtils.FormatMessageList(jumpjetFields) + "\n\n";
+
+			message += "Condition definitions have been automatically migrated.\n"
+				+ "The following definitions reference fields that must be manually moved to Locomotors:\n"
+				+ UpdateUtils.FormatMessageList(locations.Select(n => n.Item1 + " (" + n.Item2 + ")"));
+
+			if (locations.Any())
+				yield return message;
+
+			locations.Clear();
+		}
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			var addNodes = new List<MiniYamlNode>();
+			foreach (var mobileNode in actorNode.ChildrenMatching("Mobile"))
+			{
+				var checkFields = locomotorFields.Append(subterraneanFields).Append(jumpjetFields);
+				if (checkFields.Any(f => mobileNode.ChildrenMatching(f).Any()))
+					locations.Add(Tuple.Create(actorNode.Key, actorNode.Location.Filename));
+
+				var tunnelConditionNode = mobileNode.LastChildMatching("TunnelCondition");
+				if (tunnelConditionNode != null)
+				{
+					var grantNode = new MiniYamlNode("GrantConditionOnTunnelLayer", "");
+					grantNode.AddNode("Condition", tunnelConditionNode.Value.Value);
+					addNodes.Add(grantNode);
+					mobileNode.Value.Nodes.Remove(tunnelConditionNode);
```

Comment:
This applies to the entire file: You should use the respective `Remove` and `Add` methods instead of using `.Value.Nodes`.

Reply:
Done, but only In places I was confident about (this rule was pretty much written by @pchote).